lenskit-core/src/main/java/org/lenskit/data/dao/DataAccessObject.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * LensKit, an open source recommender systems toolkit.
+ * Copyright 2010-2014 LensKit Contributors.  See CONTRIBUTORS.md.
+ * Work on LensKit has been funded by the National Science Foundation under
+ * grants IIS 05-34939, 08-08692, 08-12148, and 10-17697.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 51
+ * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+package org.lenskit.data.dao;
+
+import it.unimi.dsi.fastutil.longs.LongSet;
+import org.lenskit.data.entities.Entity;
+import org.lenskit.data.entities.EntityType;
+import org.lenskit.data.entities.TypedName;
+import org.lenskit.util.IdBox;
+import org.lenskit.util.io.ObjectStream;
+
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Interface for accessing data in LensKit.
+ *
+ * LensKit data is represented by *entities*; see {@link org.lenskit.data.entities}
+ *
+ * DAOs **must** be thread-safe, but individual streams they return need not be.
+ */
+public interface DataAccessObject {
+    /**
+     * Get the types of entities that are available in this DAO.
+     * @return The set of available entity types.
+     */
+    Set<EntityType> getEntityTypes();
```

Comment:
I guess I'm OK with this. In this design. I could see this introducing some extra objects whose purpose is to be a data structure cacheing and optimizing over the DAO. An example could be a user rating count or item rating count object. In LK2 these types of "queries" could reasonably fall under the guise of a DAO, but in this model I think they become something else.